[2.0.x] PLATFORMIO.INI - TMC26xx regression, ELF improvement, re-arrange

Bob-the-Kuhn opened the pull request:
1) When compiling for DUE the new TMC26xx library gives a "_BV()" not defined compile error.  This PR reverts it to the previous one.

2) Add build flags so that the ELF file contains debugging information so that the source code can be seen along with the assembly code

3) Re-arranged the order of the environments.  Kept Mega2560  * Mega1280 at the top and moved MELZI so that everything else is in alphabetical order.

Bob-the-Kuhn committed:
TMC 26xx regression, ELF improvement

thinkyhead commented on platformio.ini:
```diff
@@ -30,12 +30,15 @@ lib_deps =
   https://github.com/lincomatic/LiquidTWI2/archive/master.zip
-  https://github.com/trinamic/TMC26XStepper/archive/master.zip
+#  https://github.com/trinamic/TMC26XStepper/archive/master.zip
+  https://github.com/trinamic/TMC26XStepper.git
```
How can this be getting something different from `master.zip`? The download links on their github page considers these to be equivalent.

Would this pull down a version that works?
https://github.com/trinamic/TMC26XStepper/archive/0.1.zip
Should we ask Trinamic to give us some more version tags?

It would be ideal if we can avoid having a dependency on `git` as this is frustrating for n00bs using Windows, which conspicuously lacks a git installation. ("Free Linus!")

thinkyhead replied:
So, I just did a compare on them and they are identical. What's up with the `_BV` missing issue, then?

Bob-the-Kuhn commented:
All I know for sure is that it's repeatable on my Win10 machine.

I'm thinking that it's not finding the TMC26xx files when https://github.com/trinamic/TMC26XStepper/archive/master.zip is enabled and https://github.com/trinamic/TMC26XStepper.git is commented out.

I get the same compile errors in steps 4 & 7 if I do the following:
1. uncomment https://github.com/trinamic/TMC26XStepper/archive/master.zip and comment out.https://github.com/trinamic/TMC26XStepper.git
2. PIO Clean(Due)
3. Delete the .piolibdeps directory
4. PIO Build(Due)
5. PIO Clean(Due)
6. Delete the contents of the directory .piolibdeps/master@src-d7635cbe3386c6c3bf6c9ab5b4141aa8 but leave the directory there
7. PIO Build(Due)

The contents of the directory .piolibdeps/master@src-d7635cbe3386c6c3bf6c9ab5b4141aa8 is NOT restored in steps 7 and 8.  

If I try to do the same thing with https://github.com/trinamic/TMC26XStepper.git enabled then the contents of  .piolibdeps\TMC26XStepper are restored.

----

Here's the error messages I get.
```
Archiving .pioenvs\DUE\lib371\libTMC2208Stepper.a
Compiling .pioenvs\DUE\lib21f\master\TMC26XStepper.cpp.o
.piolibdeps\master\TMC26XStepper.cpp: In member function 'char TMC26XStepper::getStallGuardThreshold()':
.piolibdeps\master\TMC26XStepper.cpp:375:23: error: '_BV' was not declared in this scope
if (result & _BV(6)) {
^

.piolibdeps\master\TMC26XStepper.cpp: In member function 'void TMC26XStepper::send262(long unsigned int)':
.piolibdeps\master\TMC26XStepper.cpp:956:30: error: 'SPCR' was not declared in this scope
unsigned char oldMode =  SPCR & SPI_MODE_MASK;
^
.piolibdeps\master\TMC26XStepper.cpp:956:37: error: 'SPI_MODE_MASK' was not declared in this scope
unsigned char oldMode =  SPCR & SPI_MODE_MASK;
^
```

thinkyhead commented:
The problem is that `lib_ignore` can't ignore it by name if it's fetched by URL. The solution is to move this library out of the `[common]` section and into a new section, then add `${TMC26XStepper.lib_deps}` to the `env` entries that are compatible with it, and leave it out of those which aren't.

```ini
[env:megaatmega2560]
platform    = atmelavr
framework   = arduino
board       = megaatmega2560
build_flags = ${common.build_flags}
board_f_cpu = 16000000L
lib_deps    = ${common.lib_deps}
  ${TMC26XStepper.lib_deps}
src_filter  = ${common.default_src_filter}
```

thinkyhead commented:
Hmm… Well, that almost works.

**Background:** All the declared libraries are placed in folders in the `.piolibdeps` folder and given unique names.
- Those that are in the PIO registry and can be included by name are simply named after the library. For example `Adafruit NeoPixel_ID28`.
- Those that are included as a Github archive are given the name of the archive, plus a unique hash. For example, `master@src-9f6a183b4f3ef33e94fcf23bb604c074`. (Not a commit id.)
- The exception to this is `U8glib-HAL` because it has a `library.json` file in its archive with an entry like this: `"name": "U8glib-HAL"`. So the folder gets named `U8glib-HAL_ID1932`.

Something weird happens if you put the dependency in a block other than `[common]`.
```ini
[TMC26XStepper]
lib_deps = https://github.com/trinamic/TMC26XStepper/archive/master.zip
```
The folder is just named `master` with no unique ID. As a result, this actually works to ignore it: `lib_ignore master`.

Thankfully, there is another way that is totally within our control. You can use the commit-id instead…
```
https://github.com/trinamic/TMC26XStepper/archive/c1921b4.zip
```
…and the folder is named `c1921b4`. That allows you to use `lib_ignore c1921b4`.

thinkyhead commented:
Normally the `lib_ignore  TMC26XStepper` in `[env:DUE]` filters out the `TMC26XStepper` library. The problem is that `lib_ignore` can't ignore a library by **name** if it's fetched by **URL**. But there is a way to fix it!

**Background:** All the declared libraries are placed in folders in the `.piolibdeps` folder and given unique names.
- Those that are in the PIO registry and can be included by name are simply named after the library. For example `Adafruit NeoPixel_ID28`.
- Those that are included using a Github archive URL are given the name of the archive, plus a unique hash. For example, `master@src-9f6a183b4f3ef33e94fcf23bb604c074`. (Not a commit id.)
- The exception to this is `U8glib-HAL` because it has a `library.json` file in its archive with an entry like this: `"name": "U8glib-HAL"`. So the folder gets named `U8glib-HAL_ID1932`.

Something weird happens if you put the dependency in a block other than `[common]`.
```ini
[TMC26XStepper]
lib_deps = https://github.com/trinamic/TMC26XStepper/archive/master.zip
```
The folder is just named `master` with no unique ID. As a result, this actually works to ignore it: `lib_ignore master`.

Thankfully, there is another way that is totally within our control. You can use the commit-id instead…
```
https://github.com/trinamic/TMC26XStepper/archive/c1921b4.zip
```
…and the folder is named `c1921b4`. That allows you to use `lib_ignore c1921b4`.

thinkyhead committed:
Fix lib_ignore in [env:DUE]

Use the commit id of the archive and you can then refer to it by a known name in lib_ignore.
